### PR TITLE
DOM-35418: Bug in CR provisioning and update

### PR DESCRIPTION
**Pull request by adp312**

### CODE REVIEW

File: pkg/controller/actions/actions.go [hunk 1]
```diff
@@ -31,8 +44,10 @@ func CreateOrUpdateOwnedResource(ctx *core.Context, owner metav1.Object, control
 		return nil
 	}
 
-	ctx.Log.V(1).Info("Applying patch to object", "gvk", gvk, "object", controlled, "patch", string(patchResult.Patch))
-	if err = ctx.Patch.Annotator.SetLastAppliedAnnotation(controlled); err != nil {
+	ctx.Log.V(1).Info("Applying patch to object", "gvk", gvk, "object", controlled, "patch",
+		string(patchResult.Patch))
+	err = ctx.Patch.Annotator.SetLastAppliedAnnotation(controlled)
```
Comment (adp312) [inline pkg/controller/actions/actions.go:49]: Not sure

File: pkg/controller/actions/actions.go [hunk 2]
```diff
@@ -95,34 +111,18 @@ func DeleteStorage(ctx *core.Context, opts []client.ListOption) error {
...
-		if err = ctx.Patch.Annotator.SetLastAppliedAnnotation(controlled); err != nil {
-			return
-		}
+	return gvks[0], nil
+}
 
-		ctx.Log.V(1).Info("Creating controlled object", "gvk", gvk, "object", controlled)
-		err = ctx.Client.Create(ctx, controlled)
+func createOwnedResource(ctx *core.Context, controlled client.Object) error {
+	err := ctx.Patch.Annotator.SetLastAppliedAnnotation(controlled)
```
Comment (adp312) [inline pkg/controller/actions/actions.go:123]: Not sure

File: pkg/controller/actions/actions.go [hunk 3]
```diff
@@ -1,25 +1,38 @@
...
+	if err != nil {
+		return err
+	}
+
+	gvk, err := getObjectKind(ctx, controlled)
+	if err != nil {
+		return err
+	}
+
+	found := controlled.DeepCopyObject().(client.Object)
```
Comment (ddl-kevin) [inline pkg/controller/actions/actions.go:28]: What does this syntax do? `DeepCopyObject().(client.Object)`

Is `found` the right name for this variable? It looks like we perform the actual check on the next line.

**Comment by adp312** [comment]
The lint warning is already fixed in a separate commit. I will rebase accordingly when the time comes.

File: pkg/controller/actions/actions.go (hunk 1 above)
Comment (ddl-kevin) [reply pkg/controller/actions/actions.go:49]: This isnt clear to me either, shouldnt kube-apiserver populate this field. TBH I dont understand the new apply mechanism that this field relates to :(. 

@sonnysideup ?

**Review by ddl-kevin — APPROVED** [review]
This looks reasonable. Was there a reason you went this route vs. just reverting the changes?

File: pkg/controller/actions/actions.go (hunk 3 above)
Comment (adp312) [reply pkg/controller/actions/actions.go:28]: - The deepCopy method (it's auto-generated) creates a "true" copy of a the data object including proper cloning of all nested/constituent objects. I haven't looked at the details of how this is used here specifically, but that's a normally accepted practice.
- `found` is likely not a right now. It's my copy/paste of Sonny's code (a bad excuse).

File: pkg/controller/actions/actions.go (hunk 1 above)
Comment (adp312) [reply pkg/controller/actions/actions.go:49]: Never mind, then. I will separately look at what's going on there. This entire piece of code will need some final refactoring, anyway.

**Comment by adp312** [comment]
Please note that the branch name here doesn't make sense. It should be something like "petrov.DOM-35418.service-update-error"

**Pull request merged**